src/decimal.rs

```diff
@@ -514,116 +528,168 @@ impl Decimal {
         self.round_dp(0)
     }
 
-    /// Returns a new `Decimal` number with the specified number of decimal points for fractional portion.
-    /// Rounding currently follows "Bankers Rounding" rules. e.g. 6.5 -> 6, 7.5 -> 8
+    /// Returns a new `Decimal` number with the specified number of decimal points for fractional
+    /// portion.
+    /// Rounding is performed using the provided `RoundingStrategy`
     ///
     /// # Arguments
     /// * `dp`: the number of decimal points to round to.
     ///
     /// # Example
     ///
     /// ```
-    /// use rust_decimal::Decimal;
+    /// use rust_decimal::{Decimal, RoundingStrategy};
     /// use std::str::FromStr;
     ///
-    /// let pi = Decimal::from_str("3.1415926535897932384626433832").unwrap();
-    /// assert_eq!(pi.round_dp(2).to_string(), "3.14");
+    /// let tax = Decimal::from_str("3.4395").unwrap();
+    /// assert_eq!(tax.round_dp_with_strategy(2, RoundingStrategy::RoundHalfUp).to_string(), "3.44");
     /// ```
-    pub fn round_dp(&self, dp: u32) -> Decimal {
+    pub fn round_dp_with_strategy(&self, dp: u32, strategy: RoundingStrategy) -> Decimal {
+        // Short circuit for zero
+        if self.is_zero() {
+            return Decimal {
+                lo: 0,
+                mid: 0,
+                hi: 0,
+                flags: flags(self.is_sign_negative(), dp),
+            };
+        }
 
-        let old_scale = self.scale();
+        // return early if decimal has a smaller number of fractional places than dp
+        // e.g. 2.51 rounded to 3 decimal places is 2.51
+        if self.scale() <= dp {
+            return *self
+        }
 
-        if dp < old_scale {
-            // Short circuit for zero
-            if self.is_zero() {
-                return Decimal {
-                    lo: 0,
-                    mid: 0,
-                    hi: 0,
-                    flags: flags(self.is_sign_negative(), dp),
-                };
-            }
+        let old_scale = self.scale();
```

Comment:
Any reason you do this after the previous comparison? As it is it just means the flags get evaluated twice.